lib/_http_server.js

```diff
@@ -255,9 +256,18 @@ function writeHead(statusCode, reason, obj) {
 // Docs-only deprecated: DEP0063
 ServerResponse.prototype.writeHeader = ServerResponse.prototype.writeHead;
 
+function Server(options, requestListener) {
+  if (typeof options === 'function') {
+    requestListener = options;
+    options = {};
+  } else if (options == null || typeof options === 'object') {
+    options = util._extend({}, options);
+  }
+
+  this.IncomingMessage = options.IncomingMessage || IncomingMessage;
+  this.ServerResponse = options.ServerResponse || ServerResponse;
```

Comment:
I don't have an answer to this but I would like to flag the fact that this means that `IncomingMessage` & `ServerResponse` can be altered on the fly, rather than solely through options. It's possible this isn't desirable and we should instead use Symbols, or maybe it is desirable and we should document it as such.

Reply:
I would actually prefer them to be symbols and be private, if there there are no performance regressions.

Reply:
@apapirovski @mcollina please check out my latest commit, I made them private.
I'm going to split them into http1/2 commits after you reviewed it.